Source/WebCore/platform/graphics/mac/IconMac.mm

```diff
@@ -27,11 +27,141 @@
+    }
+    CGContextRef context = CGBitmapContextCreate(nil, iconSideLength, iconSideLength, 8, 4 * iconSideLength, colorSpace, kCGImageAlphaPremultipliedLast);
+    CFRelease(colorSpace);
+
+    CGContextSetInterpolationQuality(context, kCGInterpolationHigh);
+    CGContextDrawImage(context, destinationRect, squaredImage);
+
+    CGImageRef scaledImage = CGBitmapContextCreateImage(context);
+    if (!scaledImage) {
+        CGContextRelease(context);
+        return [[NSImage alloc] initWithCGImage:squaredImage size:NSZeroSize];
+    }
+
+    NSImage *result = [[NSImage alloc] initWithCGImage:scaledImage size:NSZeroSize];
+
+    CGImageRelease(scaledImage);
+    CGContextRelease(context);
+
+    return result;
+}
+
+static NSImage* iconForImageFile(NSURL *file)
+{
+    ASSERT_ARG(file, [file isFileURL]);
+
+    NSDictionary *options = @{
+        (id)kCGImageSourceCreateThumbnailFromImageIfAbsent: @YES,
+        (id)kCGImageSourceThumbnailMaxPixelSize: @(iconSideLength),
+        (id)kCGImageSourceCreateThumbnailWithTransform: @YES,
+    };
+    RetainPtr<CGImageSource> imageSource = adoptCF(CGImageSourceCreateWithURL((CFURLRef)file, 0));
+    RetainPtr<CGImageRef> thumbnail = adoptCF(CGImageSourceCreateThumbnailAtIndex(imageSource.get(), 0, (CFDictionaryRef)options));
+    if (!thumbnail) {
+        LOG_ERROR("Error creating thumbnail image for image: %@", file);
+        return [[NSWorkspace sharedWorkspace] iconForFile:file.absoluteString];
+    }
+
+    return thumbnailSizedImageForImage(thumbnail.get());
+}
+
+static NSImage * iconForVideoFile(NSURL *file)
+{
+    ASSERT_ARG(file, [file isFileURL]);
+
+    RetainPtr<AVURLAsset> asset = adoptNS([PAL::allocAVURLAssetInstance() initWithURL:file options:nil]);
+    RetainPtr<AVAssetImageGenerator> generator = adoptNS([PAL::allocAVAssetImageGeneratorInstance() initWithAsset:asset.get()]);
+    [generator setAppliesPreferredTrackTransform:YES];
+
+    NSError *error = nil;
+    RetainPtr<CGImageRef> imageRef = adoptCF([generator copyCGImageAtTime:PAL::kCMTimeZero actualTime:nil error:&error]);
+    if (!imageRef) {
+        LOG_ERROR("Error creating image for video '%@': %@", file, error);
+        return [[NSWorkspace sharedWorkspace] iconForFile:file.absoluteString];
+    }
+
+    return thumbnailSizedImageForImage(imageRef.get());
+}
+
+static NSImage * iconForFile(NSURL *file)
+{
+    ASSERT_ARG(file, [file isFileURL]);
+
+    NSString *fileExtension = file.pathExtension;
+    if (!fileExtension.length)
+        return nil;
+
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    RetainPtr<CFStringRef> fileUTI = adoptCF(UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, (CFStringRef)fileExtension, 0));
+
+    if (UTTypeConformsTo(fileUTI.get(), kUTTypeImage))
+        return iconForImageFile(file);
+
+    if (UTTypeConformsTo(fileUTI.get(), kUTTypeMovie))
+        return iconForVideoFile(file);
+ALLOW_DEPRECATED_DECLARATIONS_END
```

Comment:
Please figure out which indentation is correct for these, and use it everywhere. Also probably merge this one with the following one.